tfe/resource_tfe_notification_configuration_test.go

```diff
@@ -0,0 +1,436 @@
+	resource.Test(t, resource.TestCase{
+		PreCheck:     func() { testAccPreCheck(t) },
+		Providers:    testAccProviders,
+		CheckDestroy: testAccCheckTFENotificationConfigurationDestroy,
+		Steps: []resource.TestStep{
+			{
+				Config:      testAccTFENotificationConfiguration_wrongDestinationType,
+				ExpectError: regexp.MustCompile(``),
+			},
+		},
+	})
+}
+
+func TestAccTFENotificationConfiguration_wrongTriggers(t *testing.T) {
+	resource.Test(t, resource.TestCase{
+		PreCheck:     func() { testAccPreCheck(t) },
+		Providers:    testAccProviders,
+		CheckDestroy: testAccCheckTFENotificationConfigurationDestroy,
+		Steps: []resource.TestStep{
+			{
+				Config:      testAccTFENotificationConfiguration_wrongTriggers,
+				ExpectError: regexp.MustCompile(``),
+			},
+		},
+	})
+}
+
+func TestAccTFENotificationConfiguration_duplicateTriggers(t *testing.T) {
+	notificationConfiguration := &tfe.NotificationConfiguration{}
+
+	resource.Test(t, resource.TestCase{
+		PreCheck:     func() { testAccPreCheck(t) },
+		Providers:    testAccProviders,
+		CheckDestroy: testAccCheckTFENotificationConfigurationDestroy,
+		Steps: []resource.TestStep{
+			{
+				Config: testAccTFENotificationConfiguration_duplicateTriggers,
+				Check: resource.ComposeTestCheckFunc(
+					testAccCheckTFENotificationConfigurationExists(
+						"tfe_notification_configuration.foobar", notificationConfiguration),
+					testAccCheckTFENotificationConfigurationAttributesDuplicateTriggers(notificationConfiguration),
+					resource.TestCheckResourceAttr(
+						"tfe_notification_configuration.foobar", "destination_type", "generic"),
+					resource.TestCheckResourceAttr(
+						"tfe_notification_configuration.foobar", "name", "notification_duplicate_triggers"),
+					// Just test the number of items in triggers
+					// Values in triggers attribute are tested by testCheckTFENotificationConfigurationAttributes
+					resource.TestCheckResourceAttr(
+						"tfe_notification_configuration.foobar", "triggers.#", "1"),
+					resource.TestCheckResourceAttr(
+						"tfe_notification_configuration.foobar", "url", "http://example.com"),
+				),
+			},
+			{
+				ResourceName:            "tfe_notification_configuration.foobar",
+				ImportState:             true,
+				ImportStateVerify:       true,
+				ImportStateVerifyIgnore: []string{"token", "workspace_external_id"},
+			},
+		},
+	})
+}
+
+func testAccCheckTFENotificationConfigurationExists(n string, notificationConfiguration *tfe.NotificationConfiguration) resource.TestCheckFunc {
+	return func(s *terraform.State) error {
+		tfeClient := testAccProvider.Meta().(*tfe.Client)
+
+		rs, ok := s.RootModule().Resources[n]
+		if !ok {
+			return fmt.Errorf("Not found: %s", n)
+		}
+
+		if rs.Primary.ID == "" {
+			return fmt.Errorf("No instance ID is set")
+		}
+
+		nc, err := tfeClient.NotificationConfigurations.Read(ctx, rs.Primary.ID)
+		if err != nil {
+			return err
+		}
+
+		*notificationConfiguration = *nc
```

Comment:
Let's try and avoid assignment of values in check functions like this that are designed to do something else - it kind of hides the intent.

Usually, if I need something like this, I add another function to do it, or I just write an inline anonymous function.

Looking at how it's used though, you could probably just move all of the logic that reads in `notificationConfiguration` into `testAccCheckTFENotificationConfigurationAttributes`, et al, as that check function will also get the state and you can fetch the ID to pass into `tfeClient.NotificationConfigurations.Read` from there. You could also write a helper to get the data if you are worried about duplication.

Reply:
@vancluever this is actually a super common use pattern within providers and is also used everywhere in this provider, so I strongly advice to keep it like this.

Reply:
@svanharmelen this does look like it's the case, sounds good.